python/cuml/dask/common/utils.py

```diff
@@ -153,12 +160,106 @@ def patch_cupy_sparse_serialization(client):
 
     client : dask.distributed.Client client to use
     """
+
+    from distributed.protocol import register_generic
+
     def patch_func():
         def serialize_mat_descriptor(m):
             return cp.cupy.cusparse.MatDescriptor.create, ()
 
+        register_generic(Base)
+
         copyreg.pickle(cp.cupy.cusparse.MatDescriptor,
                        serialize_mat_descriptor)
 
     patch_func()
     client.run(patch_func)
+
+
+class MultiHolderLock:
+    """
+    A per-process synchronization lock allowing multiple concurrent holders
```

Comment:
Let me ask a dumb question for my understanding... if everyone ran dask with `n_threads=1` as an option, would there be a need for locks?

Reply:
If everyone ran Dask with `n_threads=1` (which I believe is the default in the `LocalCUDACluster` then this lock is already bounded from above w/ 1 concurrent task incrementing `concurrent_tasks`. 

While users have the option to modify `n_threads` on the Dask cluster globally for all executions, I created this knob to provide an optional upper-bound at the level of a single execution path through the Dask graph so that users could better fine-tune the use of the memory for a specific model. 

For example, if I'm exploring a dataset and I'm just going to town with different modeling options, this allows me to increase the `n_threads` on the workers globally while being able to bound different models. I could probably set my linear models to utilize much more parallelism than I would, say, UMAP, or RF.